### ISSUE TITLE
Valid Case for: !!meta within !!meta

### ISSUE THREAD

**Issue by peter1000** [MichaelHatherly/Docile.jl#114]
I think I found a valid case to have `!!meta in !!meta`

```julia
module Lexicon

"""!!summary(*Lexicon* is a [Julia](http://www.julialang.org) package documentation generator and viewer.)

Author: !!setget(author:Michael Hatherly)
Copyright (c) !!setget(copyright:Michael Hatherly and other contributors)
License: !!setget(license:[MIT](https://github.com/MichaelHatherly/Lexicon.jl/blob/master/LICENSE.md))

Lexicon provides access to the documentation created by the `@doc` macro from
[*Docile*](https://github.com/MichaelHatherly/Docile.jl). *Lexicon* allows querying of package
documentation from the Julia REPL and building standalone documentation that can be hosted on GitHub
Pages or [Read the Docs](https://readthedocs.org/).

!!set(section:Lexicon.md/Lexicon/1)
"""
Lexicon

include.....

end
```

As I mentioned before [I would like to collect all outside hyperlinks](https://github.com/MichaelHatherly/Docile.jl/pull/110) so it's possible to write a Section with all links.

To do that I added something like my PR to Lexicon/Extensions: actually I rethought it and will change so that there is only 1 meta for `!!href` probably a dict - so that I can easily fetch it.

`!!href(KeyTxt:ValueHyperlink)`  meta: `:hrefs`

---

The above whould be rewritten like:

```julia
module Lexicon

"""!!summary(*Lexicon* is a !!href(Julia:http://www.julialang.org) package documentation generator and viewer.)

Author: !!setget(author:Michael Hatherly)
Copyright (c) !!setget(copyright:Michael Hatherly and other contributors)
License: !!setget(license:!!href(MIT:https://github.com/MichaelHatherly/Lexicon.jl/blob/master/LICENSE.md)

Lexicon provides access to the documentation created by the `@doc` macro from
!!href(Docile:https://github.com/MichaelHatherly/Docile.jl).  *Lexicon* allows querying of package documentation from the Julia REPL and building standalone documentation that can be hosted on GitHub Pages or !!href(Read the Docs:https://readthedocs.org/).

!!set(section:Lexicon.md/Lexicon/1)
"""
Lexicon

include.....

end
```

and returned we would have something like

```julia
hrefs = (
Julia  => http://www.julialang.org
Docile => https://github.com/MichaelHatherly/Docile.jl
Read the Docs => https://readthedocs.org/
MIT => https://github.com/MichaelHatherly/Lexicon.jl/blob/master/LICENSE.md
)

# Sort it and add it in the format whatever one uses.
```

Maybe better would it: as this is format unspecific add it to Docile and have each format a own `get_href` in Lexicon so one can slice it back immediately.

`!!href(MIT:https://github.com/MichaelHatherly/Lexicon.jl/blob/master/LICENSE.md)!!get_href(MIT)`

Or do all in Lexicon format specific always splicing it back without an extra `!!get_href`

---

**Anyway the main point is:** it seems to me that this is a valid case to allow and parse: `!!meta within !!meta` which probably should be done in Docile if we want it.

[::META"longform"](https://github.com/MichaelHatherly/Docile.jl/blob/master/src/Extensions/Extensions.jl#L80) does it to some extend already: maybe allow it as a default for all and allow deep nesting?

I guess in such case it would be more appropriate to do it directly in: `Formats.extractmeta!`  instead of each `Formats.metamacro`
?

**Comment by peter1000** [comment]
Probably one could use such similar to wikipedia footnotes [References](http://en.wikipedia.org/wiki/Julia_%28programming_language%29#References) at least in html.

**Comment by MichaelHatherly** [comment]
> and returned we would have something like
>
> ```
> hrefs = (
> Julia  => http://www.julialang.org
> Docile => https://github.com/MichaelHatherly/Docile.jl
> Read the Docs => https://readthedocs.org/
> MIT => https://github.com/MichaelHatherly/Lexicon.jl/blob/master/LICENSE.md
> )
> # Sort it and add it in the format whatever one uses.
> ```

Where is it being returned to?

For nested metamacros I think the one place we definitely won't have then is in `!!get`, `!!set`, and `!!setget`. `!!summary` is more of an option for nesting though.

> Or do all in Lexicon format specific always splicing it back without an extra `!!get_href`

Definitely no to the `!!get_href` in Docile. The `!!href` stuff is more applicable to Lexicon I believe.

> maybe allow it as a default for all and allow deep nesting?

If we did move to default nesting then yes it would go in `extractmeta!`, but as soon as we do that then we loose non-nesting completely. Nesting is a single line to be added to a definition if the author would like that behaviour.  You can't go the other way though and disable nesting on a per-definition basis easily (or at all perhaps).

**Comment by peter1000** [comment]
>  You can't go the other way though and disable nesting on a per-definition basis easily (or at all perhaps).

On the otherhand I think this has a much easier solution: `Just don't write nested docstrings if not wanted - as simple as that`

* A bit unlike you on my thought if we allow nested it should be a feature one can expect to work for all: the only exception would be in my mind: `!!get()` as it is somewhat different and only retrieves already added metas.

---

>> hrefs
> Where is it being returned to?

My thought was like this but a bit unsure what would be better:

* add one meta to the obj: `hrefs` and have a Dictionary with the **LinkText** as key and the **url** as value.

    * or maybe the other way round: each has plus and cons

* or the other thought was: because one obj docstring might have not many href and most of them migth not even have any to collect all in a similar meta associated with the module

* Also probably one will write a "Reference" or "Footnotes" section not per obj but mostly per module or pkg or multi-pkg documentation project.

**Comment by MichaelHatherly** [comment]
> A bit unlike you on my thought if we allow nested it should be a feature one can expect to work for all: the only exception would be in my mind: `!!get()` as it is somewhat different and only retrieves already added metas.

Fair point :), feel free to create a PR and we can discuss from there to see if it's worth changing the behaviour.

**Comment by peter1000** [comment]
> feel free to create a PR

Ok will give it a try.

**Comment by MichaelHatherly** [comment]
One point for keeping it as is is this:

```jl
function Formats.metamacro(::META"hypothetical", body, mod, obj)
    body = do_some_thing_with_the_unexpanded_form(body, mod, obj)
    body = extractmeta!(body, mod, obj) # Now handle nested `!!`.
    do_something_else(body, mod, obj)
end
```

I don't have a particular usecase yet, but I'd rather not loose that flexibility. PR though, we can discuss further there.

**Comment by peter1000** [comment]
I'm not clear why that would be lost. Maybe you can give a case to help me to put my mind around. I see it like this:

```
!!hypothetical(This is a new idea do whatever is needed)
```

Would that not only run once through: `function Formats.metamacro hypothetical` where you first use the body whole text as is and then extract: a new meta `:hypothetical` and do what ever it does.

Even if there would be nested metas like this there should be no problem or should there?

```
!!hypothetical(This is a new idea !!href(wiki:http://en.wikipedia.org/wiki/New_Idea) do whatever is needed)
```

Would that not also only run once through: `function Formats.metamacro hypothetical` where you first use the body whole text as is and then extract: a new meta `:hypothetical` - which will internally also have to extract meta: `!!href`  and after that do what ever it does.

Maybe I got it all wrong?

**Comment by MichaelHatherly** [comment]
> I'm not clear why that would be lost.

Having nesting as default is equivalent to the current pattern

```jl
function metamacro(::META"...", body, mod, obj)
    body = extractmeta!(body, mod, obj)
    ...
end
```

This means nothing can happen prior to the line ``body = extractmeta!(body, mod, obj)`` if we move to automatically nesting.

**Comment by peter1000** [comment]
I did work on this a bit and think I figured two option out.

- one is: complete nested extraction
- two is: a variation on: [
> This means nothing can happen prior to the line body = extractmeta!(body, mod, obj) if we move to automatically nesting.

where one can do something after the main !!meta is handled.

**Comment by peter1000** [comment]
that second version had some initial error with unicode so I skip that for now:

The basic idea is simple: First extract the first body (which would alow to do anything with it and than extract recursively all inner in one go.

Anyway I skip that PR for now and just write a couple of test for the other - looks good so far.

**Comment by MichaelHatherly** [comment]
Fixed in 3a735a7c288d8232d3f39a87ed3d7e3df7b69710.
